### PR TITLE
Implement additional functions of explorer

### LINKED ISSUE #20: Non-blocking loading of file structure in Explorer
### What feature do you suggest?
See the additional context.

### How will the feature be implemented?
Using threads (or `asyncio` if it is supported in `qiwis`).

### Additional context

> The running time of loading the file structure could be long since it is socket communication and the file structure might grow huge someday. Then it will block the main thread, so threading might be necessary. How do you think of this?

Yes, you're right!
Indeed, in ARTIQ, it is implemented by using `await` so it doesn't guarantee the stability of GUI.
But I think it isn't a good implementation and as you suggested, we need to implement to keep the GUI from breaking.

Then, how about extracting it as a new issue and not implementing it in this issue?
It looks not a necessary function, but an advanced function!

_Originally posted by @BECATRUE in https://github.com/snu-quiqcl/iquip/issues/19#issuecomment-1561374272_

### LINKED ISSUE #21: Lazy loading of file structure in Explorer
### What feature do you suggest?
See the additional context.

### How will the feature be implemented?
TBD.

It looks a great idea!
As you said, it seems to need some effort, so let's make a new issue.

_Originally posted by @BECATRUE in https://github.com/snu-quiqcl/iquip/issues/19#issuecomment-1561375914_

### LINKED ISSUE #28: Unnecessary pylint disable comment in `apps.explorer`
https://github.com/snu-quiqcl/iquip/blob/ba81d66d991da9a8f9d296a60df91a0d4967ecac/iquip/apps/explorer.py#L33

Since it has more than 2 public methods, this comment is not necessary.

### PR DESCRIPTION
**Pull request by BECATRUE**
I implemented the additional function of #20, #21, and resolved the Pylint issue of #28.

# Threading for calling the command line
- Just used `theading.Thread`.
- In according to an [example](https://wiki.qt.io/Updating-QML-content-from-Python-threads), it may be no problem.

# Lazy loading of file structure
- When initially loaded, directories have an empty item underneath to indicate that it is a directory.
- When the directory is expanded for the first time, `_addFile()` is called and the lower files are loaded.

Close #20, Close #21, and Close #28.

### CODE REVIEW

**Comment by kangz12345** [comment]
FYI, you can use "resolve" as a closing keyword as well:
This PR _resolves_ #28.

File: iquip/apps/explorer.py [hunk 1]
```diff
@@ -54,15 +56,36 @@ def loadFileTree(self):
...
 
     def _addFile(self, path: str, parent: Union[QTreeWidget, QTreeWidgetItem]):
-        """Searches all files in path and adds them into parent.
+        """Searches only files in path, not in deeper path and adds them into parent.
 
         A file or directory which starts with "_" will be ignored, e.g. __pycache__/.
 
         Args:
             path: The path of the directory to search experiment files.
+            parent: The widget corresponding to the path.
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: Oops, nice work :)

File: iquip/apps/explorer.py [hunk 2]
```diff
@@ -54,15 +56,36 @@ def loadFileTree(self):
...
+
+        This will be called when a directory item is expanded,
+        so it makes loading files lazy.
+
+        Args:
+            experimentFileItem: The expanded file item.
+        """
+        if experimentFileItem.childCount() != 1 or experimentFileItem.child(0).columnCount() != 0:
+            return
+        experimentFileItem.removeChild(experimentFileItem.child(0))
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: Please add a comment to explain that there is an empty item for an unloaded directory.

Comment (kangz12345) [inline iquip/apps/explorer.py]: You can use [`takeChild(0)`](https://doc.qt.io/qt-6/qtreewidgetitem.html#takeChild).

File: iquip/apps/explorer.py [hunk 3]
```diff
@@ -54,15 +56,36 @@ def loadFileTree(self):
...
+        experimentPath = self.getFullPath(experimentFileItem)
+        threading.Thread(
+            target=lambda: self._addFile(experimentPath, experimentFileItem)
+        ).start()
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: What a concise and clean code!

File: iquip/apps/explorer.py [hunk 4]
```diff
@@ -54,15 +56,36 @@ def loadFileTree(self):
...
+
+        Args:
+            experimentFileItem: The expanded file item.
+        """
+        if experimentFileItem.childCount() != 1 or experimentFileItem.child(0).columnCount() != 0:
+            return
+        experimentFileItem.removeChild(experimentFileItem.child(0))
+        experimentPath = self.getFullPath(experimentFileItem)
+        threading.Thread(
+            target=lambda: self._addFile(experimentPath, experimentFileItem)
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: When you use a `lambda` expression, you should be careful about the variable scope and the closure.
For example, I guess the current code would make no problem but in the future, the code could be modified, so the `experimentPath` or `experimentFileItem` is used after the thread started.
In that case, since the variables are _shared_, some subtle concurrency issues might occur.

Therefore, I suggest to use `args` argument in [`threading.Thread()`](https://docs.python.org/3.7/library/threading.html#threading.Thread), so the variables are _passed_ over. Or, you can use [`functools.partial()`](https://docs.python.org/3.7/library/functools.html#functools.partial) as well.

File: iquip/apps/explorer.py [hunk 5]
```diff
@@ -72,7 +95,8 @@ def _addFile(self, path: str, parent: Union[QTreeWidget, QTreeWidgetItem]):
             if experimentFile.endswith("/"):
                 experimentFileItem = QTreeWidgetItem(parent)
                 experimentFileItem.setText(0, experimentFile[:-1])
```
Comment (kangz12345) [inline iquip/apps/explorer.py:147]: Since you are now run this codes in another thread, you have to be aware of the GUI event loop in the main thread.
If you update the GUI (e.g., `setText()`) in a thread other than the main thread, it could crash.
For safety, you should separate the _working_ part and the GUI part, and the GUI part must run on the main thread.
In order to accomplish this, you may use signal-slot mechanism or [`invokeMethod()`](https://doc.qt.io/qt-6/qmetaobject.html#invokeMethod-7). You might need `QThread` for these feature (I'm not sure!).
Or you could find a nice way to use `threading.Thread` instead.

File: iquip/apps/explorer.py [hunk 6]
```diff
@@ -89,12 +113,20 @@ def openExperiment(self):
         experimentFileItem = self.explorerFrame.fileTree.currentItem()
         if experimentFileItem.childCount():
             return
-        experimentPath = experimentFileItem.text(0)
+        experimentPath = self.getFullPath(experimentFileItem)  # pylint: disable=unused-variable
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: Why is this line necessary?

File: iquip/apps/explorer.py [hunk 7]
```diff
@@ -89,12 +113,20 @@ def openExperiment(self):
...
+
+    def getFullPath(self, experimentFileItem: QTreeWidgetItem) -> str:
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: In Qt's style, I would rather say `fullPath()`.

File: iquip/apps/explorer.py [hunk 8]
```diff
@@ -89,12 +113,20 @@ def openExperiment(self):
...
+            experimentFileItem: The file item to get its full path.
+        """
+        path = experimentFileItem.text(0)
         while experimentFileItem.parent():
             experimentFileItem = experimentFileItem.parent()
-            experimentPath = posixpath.join(experimentFileItem.text(0), experimentPath)
-        experimentPath = posixpath.join(self.repositoryPath, experimentPath)
-
+            path = posixpath.join(experimentFileItem.text(0), path)
+        path = posixpath.join(self.repositoryPath, path)
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: How about using a `list` and extending it, and calling `join()` at the end just once?
I think `join` operation is more expansive since it repeatedly creates and copies strings.

File: setup.json [hunk 9]
```diff
@@ -6,7 +6,7 @@
             "show": true,
             "pos": "left",
             "args": {
-                "masterPath": "/home/rabi/artiq/"
+                "masterPath": "/home/jaehun/quiqcl/artiq/examples/"
```
Comment (kangz12345) [inline setup.json]: Maybe this is mistakenly staged here.
You can make another setup file (e.g. "setup_jaehun.json") and ignore it in your git repo.

**Review by kangz12345 — CHANGES_REQUESTED** [review]

**Review by kangz12345 — CHANGES_REQUESTED** [review]
In addition, consider the case where the user rapidly clicks a directory item several times, occurring the `itemExpanded` signal is emitted multiple times before the first thread is finished.

File: setup.json (hunk 9 above)
Comment (BECATRUE) [reply setup.json]: Oh, I made a mistake..
Thanks for offering a solution.

File: iquip/apps/explorer.py (hunk 2 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: Oh, it looks simpler!

File: iquip/apps/explorer.py (hunk 4 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: I think you told me about this issue before in the qiwis project.
I will fix it.

File: iquip/apps/explorer.py (hunk 6 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: It is a preparation step for opening the experiment builder.
For submitting or getting the arguments, we need to know the full path, not only the file name.
So I've implemented up to here first.

File: iquip/apps/explorer.py (hunk 8 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: Oh, it's better! I will apply it.

File: iquip/apps/explorer.py (hunk 5 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py:147]: I found out a few solutions of setting a callback method, but they are a little complicated.
So I decide to replace it to `QThread`.

**Comment by BECATRUE** [comment]
> In addition, consider the case where the user rapidly clicks a directory item several times, occurring the `itemExpanded` signal is emitted multiple times before the first thread is finished.

Oh, it's a keen analysis.

In the current system, however, we cannot clicks a directory item several times during the threading operation because before starting the thread, the child element is removed and the directory item has no child, so the toggle button would be removed.
Thus, I think we don't have to consider that situation.

What do you think of it?

**Comment by kangz12345** [comment]
> In the current system, however, we cannot clicks a directory item several times during the threading operation because before starting the thread, the child element is removed and the directory item has no child, so the toggle button would be removed.
> 
> Thus, I think we don't have to consider that situation.
> 
> 
> 
> What do you think of it?

Oh, it makes sense. If it is confirmed that the toggle button is removed, then it would be okay!

File: iquip/apps/explorer.py (hunk 6 above)
Comment (kangz12345) [reply iquip/apps/explorer.py]: In that case, please leave a TODO comment describing what is this variable for and how and when it will be used.

Comment (BECATRUE) [reply iquip/apps/explorer.py]: Umm.. But there is already a TODO comment in the docstring of the method, so I think an additional comment is not necessary.
How about remaining it as it is?

Comment (kangz12345) [reply iquip/apps/explorer.py]: Oh, I did not see that. Then it seems alright!

**Comment by kangz12345** [comment]
Please re-request the review after applying all the previous reviews.

**Comment by BECATRUE** [comment]
> Please re-request the review after applying all the previous reviews.

Oh, I pushed the updated version before, but I didn't notify it..
It is already updated!

**Comment by kangz12345** [comment]
You can _re-request_ a review in the _Reviewers_ section, as I just did! :)

File: iquip/apps/explorer.py [hunk 10]
```diff
@@ -31,6 +30,45 @@ def __init__(self, parent: Optional[QWidget] = None):
...
 
 
+class FileFinderThread(QThread):
+    """QThread for finding the file list using a command line.
+
+    Signals:
+        finished(experimentList, parent): Fetching the file list is finished.
+
+    Attributes:
+        path: The path of the directory to search experiment files.
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: As far as I know:
Usually we use "for", e.g., "search for experiment files".
When we use "search A" without "for", then A is the place or object where the interested targets are in, e.g., "search the area for clues".

Please let me know if it has other usage!

File: iquip/apps/explorer.py [hunk 11]
```diff
@@ -31,6 +30,45 @@ def __init__(self, parent: Optional[QWidget] = None):
...
+        parent: Union[QTreeWidget, QTreeWidgetItem],
+        callback: Callable[[List[str], Union[QTreeWidget, QTreeWidgetItem]], None]
+    ):
+        """
+        Args:
+            callback: The callback method called after this thread is finished.
+        """
+        super().__init__()
+        self.path = path
+        self.parent = parent
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: I think this will overwrite the method [`QObject.parent()`](https://doc.qt.io/qt-5/qobject.html#parent), since `QThread` inherits `QObject`.

Please change its name to, e.g., `widget`, etc.

File: iquip/apps/explorer.py [hunk 12]
```diff
@@ -31,6 +30,45 @@ def __init__(self, parent: Optional[QWidget] = None):
...
+
+    finished = pyqtSignal(list, object)
+
+    def __init__(
+        self,
+        path: str,
+        parent: Union[QTreeWidget, QTreeWidgetItem],
+        callback: Callable[[List[str], Union[QTreeWidget, QTreeWidgetItem]], None]
+    ):
+        """
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: Since this overrides the `QThread.__init__()` method and calls `super().__init__()`, please write "Extended.".

Comment (kangz12345) [reply iquip/apps/explorer.py]: In addition, write the docstrings for `path` and `parent`, since they are not in `QThread.__init__()` (`parent` seems confusing here as well).
You can just say "See the attributes section in FileFinderThread".

File: iquip/apps/explorer.py [hunk 13]
```diff
@@ -31,6 +30,45 @@ def __init__(self, parent: Optional[QWidget] = None):
...
+        Args:
+            callback: The callback method called after this thread is finished.
+        """
+        super().__init__()
+        self.path = path
+        self.parent = parent
+        self.finished.connect(callback)
+
+    def run(self):
```
Comment (kangz12345) [inline iquip/apps/explorer.py:64]: You are overriding [`QThread.run()`](https://doc.qt.io/qt-5/qthread.html#run), so please write "Overridden.".

File: iquip/apps/explorer.py [hunk 14]
```diff
@@ -31,6 +30,45 @@ def __init__(self, parent: Optional[QWidget] = None):
...
+        parent: The widget corresponding to the path.
+    """
+
+    finished = pyqtSignal(list, object)
+
+    def __init__(
+        self,
+        path: str,
+        parent: Union[QTreeWidget, QTreeWidgetItem],
+        callback: Callable[[List[str], Union[QTreeWidget, QTreeWidgetItem]], None]
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: I think the return type could be `Any`.

File: iquip/apps/explorer.py [hunk 15]
```diff
@@ -31,6 +30,45 @@ def __init__(self, parent: Optional[QWidget] = None):
...
+    ):
+        """
+        Args:
+            callback: The callback method called after this thread is finished.
+        """
+        super().__init__()
+        self.path = path
+        self.parent = parent
+        self.finished.connect(callback)
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: Is there a reason to connect the callback here in `FileFinderThread`? I think it can be connected outside, i.e., by the _user_ of this class.

File: iquip/apps/explorer.py [hunk 16]
```diff
@@ -31,6 +30,45 @@ def __init__(self, parent: Optional[QWidget] = None):
...
+
+    Signals:
+        finished(experimentList, parent): Fetching the file list is finished.
+
+    Attributes:
+        path: The path of the directory to search experiment files.
+        parent: The widget corresponding to the path.
+    """
+
+    finished = pyqtSignal(list, object)
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: There is already a signal with the same name: [`QThread.finished()`](https://doc.qt.io/qt-5/qthread.html#finished)

File: iquip/apps/explorer.py [hunk 17]
```diff
@@ -56,9 +94,12 @@ def loadFileTree(self):
...
         """
         self.explorerFrame.fileTree.clear()
-        threading.Thread(
-            target=lambda: self._addFile(self.repositoryPath, self.explorerFrame.fileTree)
-        ).start()
+        self.thread = FileFinderThread(
+            self.repositoryPath,
+            self.explorerFrame.fileTree,
+            self._addFile
+        )
```
Comment (kangz12345) [inline iquip/apps/explorer.py:107]: Please add the `parent=self` argument for a `QObject`.
See also: https://doc.qt.io/qt-5/qthread.html#details

File: iquip/apps/explorer.py [hunk 18]
```diff
@@ -113,19 +156,20 @@ def openExperiment(self):
         experimentFileItem = self.explorerFrame.fileTree.currentItem()
         if experimentFileItem.childCount():
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: Hmm.. this becomes now a problem, since during a directory is loaded, the widget item will have no child.

File: iquip/apps/explorer.py [hunk 19]
```diff
@@ -113,19 +156,20 @@ def openExperiment(self):
...
-        path = experimentFileItem.text(0)
+        path = [experimentFileItem.text(0)]
         while experimentFileItem.parent():
             experimentFileItem = experimentFileItem.parent()
-            path = posixpath.join(experimentFileItem.text(0), path)
-        path = posixpath.join(self.repositoryPath, path)
+            path.append(experimentFileItem.text(0))
+        path.append(self.repositoryPath)
+        path = posixpath.join(*reversed(path))
         return path
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: Actually I prefer not to bind a new value of different types to a variable 😅
Could you just return it directly or change the variable name?
For example:
```python
return posixpath.join(*reversed(path))
```
or
```python
paths = [...]
...
path = ...
return path
```

**Review by kangz12345 — CHANGES_REQUESTED** [review]

**Comment by kangz12345** [comment]
Oh, and GitHub only detects A if you write "This closes A, B and C".
See also: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

**Comment by BECATRUE** [comment]
> Oh, and GitHub only detects A if you write "This closes A, B and C". See also: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

Wow, thanks for notifying it! I will update the PR description.

File: iquip/apps/explorer.py (hunk 10 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: Oh, I looked it up and I think you're right!
I'm not good at English.. so please understand me😂

File: iquip/apps/explorer.py (hunk 14 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: Currently, the `callback` is always set as `_addFile()` and it does not return anything, so I set the return type as `None`.
Nevertheless, if you still want to fix it, I will do so.

File: iquip/apps/explorer.py (hunk 15 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: I thought about this part, too.
The reason why I did so is that `FileFinderThread` instances are created twice and I wanted to remove the duplicated parts as much as possible.

File: iquip/apps/explorer.py (hunk 16 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: Yes! Actually, I intended it.

Probably, the `finished` signal was created with the aim of announcing the end of the thread.
I needed the similar signal which has additional arguments, so I named it the same.

But, if you want to remain the original signal, I will rename it lik `fetched`.

File: iquip/apps/explorer.py (hunk 17 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py:107]: Oh, I was completely missing it. I will add the argument.

File: iquip/apps/explorer.py (hunk 19 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: Ah😃 Then, I will return it directly.
Also, I think it would be better to change the former `path` to `paths` since it contains lots of the paths (even if parts).

File: iquip/apps/explorer.py (hunk 18 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: Umm... it is a critical issue.

Then, how about not checking at all?
Actually, even if we completely check that the selected element is a file, we should re-check whether the file is an ARTIQ experiment file later.
This will be processed through an ARTIQ command line (probably, customed), so why don't we postpone it to this time?

Comment (BECATRUE) [reply iquip/apps/explorer.py]: If you want to check it in this branch, we can do this through comparing argumetns of `self.thread` to `experimentFileItem`.

File: iquip/apps/explorer.py (hunk 14 above)
Comment (kangz12345) [reply iquip/apps/explorer.py]: Oh, if it is only for the internal use, it would be okay. However, since you named it as a public class, I think it should be considered as a part of the public API.
For the public API, please make it general as possible!

Or you can make it an internal class `_FileFinderThread`.

File: iquip/apps/explorer.py (hunk 15 above)
Comment (kangz12345) [reply iquip/apps/explorer.py]: Oh I see.
A better design would be separating the `FileFinderThread` instantiation part as a method and calling it twice, not doing it in the `FileFinderThread`.

However, if you think it is only for the internal use, you can name it e.g. `_FileFinderThread` to tell that it is an internal class.

File: iquip/apps/explorer.py (hunk 16 above)
Comment (kangz12345) [reply iquip/apps/explorer.py]: Hmm, however, `finished` signal is emitted when a `QThread` finishes executing, after stopping the event loop.
I am not sure if it is safe to overwrite such a private signal which already has its own role (since Python does not support overloading).
If we don't have to name it `finished`, I would prefer to use another name!

File: iquip/apps/explorer.py (hunk 18 above)
Comment (kangz12345) [reply iquip/apps/explorer.py]: As you said, later we have to check more things. I think it is fine not to check at all for now!

File: iquip/apps/explorer.py (hunk 14 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: I intended to use the thread class only internally, so I will make it private.

File: iquip/apps/explorer.py (hunk 16 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: Okay! Then, I will change the name into another one.

File: iquip/apps/explorer.py [hunk 20]
```diff
@@ -30,43 +30,48 @@ def __init__(self, parent: Optional[QWidget] = None):
...
+        widget: Union[QTreeWidget, QTreeWidgetItem],
+        callback: Callable[[List[str], Union[QTreeWidget, QTreeWidgetItem]], None],
+        parent: Optional[QObject] = None
     ):
-        """
+        """Extended.
+
         Args:
             callback: The callback method called after this thread is finished.
+            See the attributes section in _FileFinderThread.
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: To be clear, please write it like:
```python
"""
...
path, widget: See the ...
callback: ...
"""

File: iquip/apps/explorer.py [hunk 21]
```diff
@@ -130,7 +137,7 @@ def _addFile(self, experimentList: List[str], parent: Union[QTreeWidget, QTreeWi
 
         Args:
             experimentList: The list of files under the parent path.
-            parent: See FileFinderThread class.
+            parent: See _FileFinderThread class.
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: Since we changed `parent` -> `widget`, it should be applied here as well.

File: iquip/apps/explorer.py [hunk 22]
```diff
@@ -30,43 +30,48 @@ def __init__(self, parent: Optional[QWidget] = None):
...
             callback: The callback method called after this thread is finished.
+            See the attributes section in _FileFinderThread.
         """
-        super().__init__()
+        super().__init__(parent=parent)
         self.path = path
-        self.parent = parent
-        self.finished.connect(callback)
+        self.widget = widget
+        self.fetched.connect(callback)
```
Comment (kangz12345) [inline iquip/apps/explorer.py]: To be clear, let's give an explicit connection option:
```suggestion
        self.fetched.connect(callback, type=Qt.QueuedConnection)
```

**Review by kangz12345 — CHANGES_REQUESTED** [review]

File: iquip/apps/explorer.py (hunk 21 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: Oh, I thought remaining it as `parent` doesn't make a confusion in this method, so I left it as it is.
But if you want to apply it, I will do that.

File: iquip/apps/explorer.py (hunk 22 above)
Comment (BECATRUE) [reply iquip/apps/explorer.py]: It seems to show our purpose well!

File: iquip/apps/explorer.py (hunk 21 above)
Comment (kangz12345) [reply iquip/apps/explorer.py]: It makes sense, but you should write "See _FileFinderThread.widget." since its `parent` refers to a different thing.
However, now you changed the argument name, so it doesn't matter!

**Review by kangz12345 — APPROVED** [review]
That was a lot of work, cheers!
